src/link/http/parseAndCheckHttpResponse.ts

```diff
@@ -51,43 +51,58 @@ async function* consumeMultipartBody(
   );
 
-  const reader = response.body.getReader();
+  const stream = response.body;
+  const reader = stream.getReader();
   let done = false;
+  let encounteredBoundary = false;
   let value: Uint8Array<ArrayBufferLike> | string | undefined;
 
-  while (!done) {
-    ({ value, done } = await reader.read());
-    const chunk = typeof value === "string" ? value : decoder.decode(value);
-    const searchFrom = buffer.length - boundary.length + 1;
-    buffer += chunk;
-    let bi = buffer.indexOf(boundary, searchFrom);
-
-    while (bi > -1) {
-      let message: string;
-      [message, buffer] = [
-        buffer.slice(0, bi),
-        buffer.slice(bi + boundary.length),
-      ];
-      const i = message.indexOf("\r\n\r\n");
-      const headers = parseHeaders(message.slice(0, i));
-      const contentType = headers["content-type"];
-      if (
-        contentType &&
-        contentType.toLowerCase().indexOf("application/json") === -1
-      ) {
-        throw new Error(
-          "Unsupported patch content type: application/json is required."
-        );
-      }
-      // nb: Technically you'd want to slice off the beginning "\r\n" but since
-      // this is going to be `JSON.parse`d there is no need.
-      const body = message.slice(i);
+  // check to see if we received the final boundary, which is a normal boundary followed by "--"
+  // as described in https://www.rfc-editor.org/rfc/rfc2046#section-5.1.1
+  const passedFinalBoundary = () =>
+    encounteredBoundary && buffer[0] == "-" && buffer[1] == "-";
 
-      if (body) {
-        yield body;
+  try {
+    while (!done) {
+      ({ value, done } = await reader.read());
+      const chunk = typeof value === "string" ? value : decoder.decode(value);
+      const searchFrom = buffer.length - boundary.length + 1;
+      buffer += chunk;
+      let bi = buffer.indexOf(boundary, searchFrom);
+      while (bi > -1 && !passedFinalBoundary()) {
+        encounteredBoundary = true;
+        let message: string;
+        [message, buffer] = [
+          buffer.slice(0, bi),
+          buffer.slice(bi + boundary.length),
+        ];
+        const i = message.indexOf("\r\n\r\n");
+        const headers = parseHeaders(message.slice(0, i));
+        const contentType = headers["content-type"];
+        if (
+          contentType &&
+          contentType.toLowerCase().indexOf("application/json") === -1
+        ) {
+          throw new Error(
+            "Unsupported patch content type: application/json is required."
+          );
+        }
+        // nb: Technically you'd want to slice off the beginning "\r\n" but since
+        // this is going to be `JSON.parse`d there is no need.
+        const body = message.slice(i);
+
+        if (body) {
+          yield body;
+        }
+        bi = buffer.indexOf(boundary);
+      }
+      if (passedFinalBoundary()) {
+        return;
       }
-      bi = buffer.indexOf(boundary);
     }
+    throw new Error("Unexpected end of multipart data");
```

Comment:
There is no perfect error message here, but some examples:

https://github.com/mscdex/dicer/blob/be552b42d8104bc6d74be03bebba20f08854259c/lib/Dicer.js#L68

https://github.com/mjackson/remix-the-web/blob/1258697c35b17830a8ac84aedd206e461fc51edd/packages/multipart-parser/src/lib/multipart.ts#L304-L305

Reply:
FWIW, Apollo Kotlin has `premature end of multipart body` there. No opinion at all, both work for me. This is just a FYI.

https://github.com/apollographql/apollo-kotlin/blob/8d37c6119a43a620be175841050913c246ec1139/tests/multipart/src/commonTest/kotlin/test/MultipartSubscriptionsMockServerTest.kt#L151

Reply:
```suggestion
    throw new Error("premature end of multipart body");
```

Being consistent with the other clients seems like a very good reason to change to that here :)